### DISCUSSION TITLE
How would one implement token-based authentication?

### DISCUSSION THREAD

**Discussion by mainrs** [maxcountryman/axum-login#85] [Q&A] [answered]
Each user has a primary account and normal username/password login. But I want them to be able to generate API keys and be able to use those for API calls.

Can I use this library to protect my routes and make them only accessible if such a token is present?

**Comment by czocher** [comment, marked as answer]
You can possibly create two `axum_login::extractors::AuthContext`s, one for password-based login, one for API key login. You'll need two different implementations of the `AuthUser` trait, one which would select the password hash, one which would select the hashed token as the result from the `get_password_hash` function.

This would allow a user to have both a normal password as well as a single token for API calls.

This is a good case for an example. I created an issue to make an example implementation of this case: https://github.com/maxcountryman/axum-login/issues/86

I implemented the example as a PR, see: https://github.com/maxcountryman/axum-login/pull/87
